lib/awful/hotkeys_popup/widget.lua

```diff
@@ -285,6 +311,52 @@ function widget.new(args)
     end
 
 
+    function widget_instance:_abbreviate_hotkeys(target)
+        -- This is a very crude method and only useful for keygroups.
+        -- It compares the current key value without modifiers, as it would be
+        -- printed (e.g 2/3/4/5/6) with the string that the entire
+        -- keygroup would display (e.g. 1/2/3/4/5/6/7/8/9/0). Then it prints
+        -- the first and the last match, separated by an ellipsis (2…6)
+        local clone = {}
+        for group, keys in pairs(target) do
+            clone[group] = gtable.map(function(value)
+                local _, slashes = string.gsub(value.key, "/", "")
+                local keygroup
+                -- Only rows of more than four keys are abbreviated.
+                -- This prevents:
+                --  - senseless abbreviations of three or two keys,
+                --  - accidental matches between a single letter and a keygroup,
+                --  - and confusing abbreviations of arrow keys.
+                if slashes > 3 then
+                    keygroup = gtable.find_first_key(self._keygroups, function(k, v)
+                        first = string.find(v.text, value.key, 1, true)
+                        return type(first) == "number"
+                    end)
+                end
+                if keygroup then
+                    local first, last
+                    for case in string.gmatch(value.key, "[^/]+") do
+                        if not first then first = case end
+                        last = case
+                    end
+                    return {
+                        key = first .. "…" .. last,
+                        mod = value.mod,
+                        description = value.description
+                    }
+                else
+                    return {
+                        key = value.key,
+                        mod = value.mod,
+                        description = value.description
+                    }
+                end
+            end, keys)
+        end
+        self._cached_awful_keys = clone
```

Comment:
why it taking `keys=self._cached_awful_keys` as input argument if anyway further here it's explicitly saving the result to `self._cached_awful_keys`?

Reply:
There's no real reason for this, I admit. I copied a subroutine of yours that uses an input argument, but couldn't find a way to get the variables to update until I saved explicitly to `self._cached_awful_keys`. Then I forgot about the input argument.

I've done a lot of tweaks in other computer with `rc.lua`, but I have very little experience in actual Lua programming.

Reply:
this should be fixed before it will be fine for merging